src/SFA.DAS.EmployerApprenticeshipsService.Application/Commands/RefreshEmployerLevyData/RefreshEmployerLevyDataCommandHandler.cs

```diff
@@ -51,50 +49,57 @@ public class RefreshEmployerLevyDataCommandHandler : AsyncRequestHandler<Refresh
+
+                declarations = declarations.Where(x => !existingIdsLookup.Contains(x.Id)).ToList();
+
+                declarations = declarations.Where(x => !DoesSubmissionPreDateTheLevy(x)).ToList();
+
+                declarations = declarations.Where(x => !IsSubmissionForFuturePeriod(x)).ToList();
+
+                var noPaymentForPeriodDeclarations = declarations.Where(x => x.NoPaymentForPeriod).ToList();
+
+                foreach (var dasDeclaration in noPaymentForPeriodDeclarations)
                 {
-                    var declaration = await _dasLevyRepository.GetEmployerDeclaration(dasDeclaration.Id, employerLevyData.EmpRef);
+                    await GetLevyFromPreviousSubmission(employerLevyData, dasDeclaration);
+                }
 
-                    if (declaration == null)
-                    {
+                var endOfYearAdjustmentDeclarations = declarations.Where(IsEndOfYearAdjustment).ToList();
 
-                        if (DoesSubmissionPreDateTheLevy(dasDeclaration))
-                        {
-                            continue;
-                        }
-
-                        if (IsSubmissionForFuturePeriod(dasDeclaration))
-                        {
-                            continue;
-                        }
-
-                        if (dasDeclaration.NoPaymentForPeriod)
-                        {
-                            await GetLevyFromPreviousSubmission(employerLevyData, dasDeclaration);
-                        }
-
-                        if (IsEndOfYearAdjustment(dasDeclaration))
-                        {
-                            await UpdateEndOfYearAdjustment(employerLevyData, dasDeclaration);
-                        }
-                        
-                        await _dasLevyRepository.CreateEmployerDeclaration(dasDeclaration, employerLevyData.EmpRef, message.AccountId);
-                        savedDeclarations.Add(dasDeclaration);
-                    }
+                foreach (var dasDeclaration in endOfYearAdjustmentDeclarations)
+                {
+                    await UpdateEndOfYearAdjustment(employerLevyData, dasDeclaration);
                 }
+
+                if (!declarations.Any()) continue;
+
+                await _dasLevyRepository.CreateEmployerDeclarations(declarations, employerLevyData.EmpRef,
+                    message.AccountId);
+
+                savedDeclarations.AddRange(declarations);
             }
 
             if (savedDeclarations.Any())
             {
-                await _mediator.PublishAsync(new ProcessDeclarationsEvent());
+                foreach (var levyData in message.EmployerLevyData)
+                {
+                    await _mediator.PublishAsync(new ProcessDeclarationsEvent
+                    {
+                        AccountId = message.AccountId,
+                        EmpRef = levyData.EmpRef
+                    });
+                }
```

Comment:
Will this loop publish more messages than necessary?  It's possible not all of the items in EmployerLevyData will have resulted in new declarations being saved.